src/next.ts

```diff
@@ -0,0 +1,64 @@
+import areInputsEqual from './are-inputs-equal';
+
+// Using ReadonlyArray<T> rather than readonly T as it works with TS v3
+export type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;
+
+type Cache<TResult> = {
+  lastThis: unknown;
+  lastArgs: unknown[];
+  lastResult: TResult;
+};
+
+type MemoizedFn<T extends (this: any, ...args: any[]) => any> = {
+  name: string;
+  clear: () => void;
+  (...args: Parameters<T>): ReturnType<T>;
+};
+
+function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
+  resultFn: TFunc,
+  isEqual: EqualityFn = areInputsEqual,
+): MemoizedFn<TFunc> {
+  let map = new WeakMap<Record<string, any>, Cache<ReturnType<TFunc>>>();
+  let key = {};
```

Comment:
I'm not sure if this is actually doing what I want with the weakmap

Ideally, it would be great for `lastThis` (and the other cache values) to be able to be garbage collected. I've tried something here with weakmap, but I might just go back to a map